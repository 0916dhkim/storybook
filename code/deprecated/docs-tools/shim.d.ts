export * from 'storybook/docs-tools';
export type * from 'storybook/docs-tools';
