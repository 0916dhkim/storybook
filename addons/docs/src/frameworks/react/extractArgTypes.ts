import { ArgTypes } from '@storybook/api';
import { PropDef, ArgTypesExtractor } from '../../lib/docgen';
import { extractProps } from './extractProps';

export const extractArgTypes: ArgTypesExtractor = (component) => {
  if (component) {
    const { rows } = extractProps(component);
    if (rows) {
      return rows.reduce((acc: ArgTypes, row: PropDef) => {
        const { type, sbType, defaultValue: defaultSummary, jsDocTags, required } = row;
        let defaultValue;
        const defaultValueString =
          defaultSummary && (defaultSummary.detail || defaultSummary.summary);
        try {
          if (defaultValueString) {
            // eslint-disable-next-line no-new-func
            defaultValue = Function(`"use strict";return (${defaultValueString})`)();
          }
          // eslint-disable-next-line no-empty
        } catch {}

        acc[row.name] = {
          ...row,
          defaultValue,
          type: { required, ...sbType },
          table: {
            type,
            jsDocTags,
            defaultValue: defaultSummary,
          },
        };
        return acc;
      }, {});
    }
  }

  return null;
};
