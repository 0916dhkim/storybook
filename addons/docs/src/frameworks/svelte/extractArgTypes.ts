/* eslint-disable new-cap */
/* eslint-disable no-underscore-dangle */
import { ArgTypes } from '@storybook/api';

import { ArgTypesExtractor, hasDocgen, extractComponentProps } from '../../lib/docgen';

type ComponentWithDocgen = {
  __docgen: {
    components: [];
    computed: [];
    data: [
      {
        defaultValue: any;
        description: string;
        keywords: [];
        kind: string;
        name: string;
        readonly: boolean;
        static: boolean;
        type: { kind: string; text: string; type: string };
        visibility: string;
      }
    ];
    description: null;
    events: [];
    keywords: [];
    methods: [];
    name: null;
    refs: [];
    slots: [];
    version: number;
  };
};

export const extractArgTypes: ArgTypesExtractor = (component) => {
  const comp: ComponentWithDocgen = new component({ props: {} });
  const docs = comp.__docgen;
  const results: ArgTypes = {};
  docs.data.forEach((item) => {
    results[item.name] = {
      control: { type: parseType(item.type.type) },
      name: item.name,
      description: item.description,
      type: {},
      defaultValue: item.defaultValue,
      table: {
        defaultValue: {
          summary: item.defaultValue,
        },
      },
    };
  });
  return results;
};

/**
 * Function to convert the type from sveltedoc-parser to a storybook type
 * @param typeName
 *  @returns string
 */
const parseType = (typeName: string) => {
  if (typeName === 'string') return 'text';
  return typeName;
};
