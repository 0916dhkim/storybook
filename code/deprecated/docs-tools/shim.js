module.exports = require('storybook/docs-tools');
