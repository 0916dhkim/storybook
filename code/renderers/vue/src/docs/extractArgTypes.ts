import type {
  SBEnumType,
  SBObjectType,
  SBScalarType,
  SBType,
  StrictArgTypes,
} from '@storybook/types';
import type { ArgTypesExtractor, DocgenInfo, PropDef } from '@storybook/docs-tools';
import { hasDocgen, extractComponentProps } from '@storybook/docs-tools';

type Schema = { kind: string; schema: [] | object; type?: string } | string;
type MetaDocgenInfo = DocgenInfo & {
  type: string;
  default: string;
  global: boolean;
  name: string;
  schema: Schema;
  tags: { name: string; text: string }[];
};

const SECTIONS = ['props', 'events', 'slots', 'methods'];

/**
 * As @enum tag is not implemented in vuedocgen, infers propdef enum type
 * from the presence of @values tag.
 */
function inferEnum(propDef: PropDef, docgenInfo: DocgenInfo): false | PropDef {
  // cast as any, since "values" doesn't exist in DocgenInfo type
  const { type, values } = docgenInfo as any;
  const matched = Array.isArray(values) && values.length && type && type.name !== 'enum';

  if (!matched) {
    return false;
  }

  const enumString = values.join(', ');
  let { summary } = propDef.type;
  summary = summary ? `${summary}: ${enumString}` : enumString;

  Object.assign(propDef.type, {
    ...propDef.type,
    name: 'enum',
    value: values,
    summary,
  });
  return propDef;
}

/**
 * @returns {Array} result
 * @returns {PropDef} result.def - propDef
 * @returns {boolean} result.isChanged - flag whether propDef is mutated or not.
 *  this is needed to prevent sbType from performing convert(docgenInfo).
 */
function verifyPropDef(propDef: PropDef, docgenInfo: DocgenInfo): [PropDef, boolean] {
  let def = propDef;
  let isChanged = false;

  // another callback can be added here.
  // callback is mutually exclusive from each other.
  const callbacks = [inferEnum];
  for (let i = 0, len = callbacks.length; i < len; i += 1) {
    const matched = callbacks[i](propDef, docgenInfo);
    if (matched) {
      def = matched;
      isChanged = true;
    }
  }

  return [def, isChanged];
}

export const extractArgTypes: ArgTypesExtractor = (component) => {
  if (!hasDocgen(component)) {
    return null;
  }
  const argTypes: StrictArgTypes = {};
  SECTIONS.forEach((section) => {
    const props = extractComponentProps(component, section);
    console.log('props', props);
    props.forEach(({ propDef, docgenInfo, jsDocTags }) => {
      const {
        name,
        type,
        description,
        default: defaultSummary,
        required,
        tags = [],
        global,
        schema,
      } = docgenInfo as MetaDocgenInfo;

      if (argTypes[name] || global) {
        return; // skip duplicate and global props
      }

      // const [result, isPropDefChanged] = verifyPropDef(propDef, docgenInfo);
      // const { name, type, description, defaultValue: defaultSummary, required } = result;

      const sbType = section === 'props' ? convert(docgenInfo as MetaDocgenInfo) : { name: 'void' };
      const nestedTypes =
        sbType.name === 'object' && section === 'props' ? nestedInfo(sbType as SBObjectType) : '';
      const definedTypes = `${type.replace(' | undefined', '')}`;
      const descriptions = `${
        tags.length ? `${tags.map((tag) => `@${tag.name}: ${tag.text}`).join('<br>')}<br><br>` : ''
      }${description} ${nestedTypes} `;

      argTypes[name] = {
        name,
        description: descriptions.replace('undefined', ''),
        defaultValue: { summary: defaultSummary },
        type: { required, ...sbType } as SBType,
        table: {
          type: { summary: definedTypes },
          jsDocTags: tags,
          defaultValue: { summary: defaultSummary },
          category: section,
        },
      };
    });
  });
  return argTypes;
};

export const convert = ({ schema: schemaType }: MetaDocgenInfo): SBType => {
  if (typeof schemaType !== 'object') {
    return { name: schemaType } as SBScalarType;
  }
  if (
    typeof schemaType === 'object' &&
    schemaType.kind === 'enum' &&
    Array.isArray(schemaType.schema)
  ) {
    const values = schemaType.schema
      .filter((item) => item !== 'undefined' && item !== null)
      .map((item) => (typeof item === 'string' ? item.replace(/"/g, '') : item));

    const sbType: SBType = { name: 'enum', value: values };
    const stringIndex = values.indexOf('string');
    const numberIndex = values.indexOf('number');
    const booleanIndex = values.indexOf('boolean');
    if (stringIndex !== -1 || numberIndex !== -1 || booleanIndex !== -1) {
      const typeName = values[stringIndex ?? numberIndex ?? booleanIndex];
      return { ...sbType, name: typeName, value: undefined } as SBScalarType;
    }
    const hasObject = values.find((item) => typeof item === 'object');

    return {
      ...sbType,
      name: hasObject ? 'array' : 'enum',
      value: hasObject ? undefined : values,
    } as SBEnumType;
  }
  if (
    typeof schemaType === 'object' &&
    schemaType.kind === 'object' &&
    typeof schemaType.schema === 'object'
  ) {
    const schemaObject = schemaType.schema as { [key: string]: MetaDocgenInfo };
    const props = Object.fromEntries(
      Object.entries(schemaObject).map(([key, value]) => {
        return [key, value as MetaDocgenInfo];
      })
    );
    return {
      name: 'object',
      value: props,
    } as SBObjectType;
  }
  return { name: schemaType } as unknown as SBType;
};

function nestedInfo(sbType: SBObjectType) {
  return Object.keys(sbType.value)
    .map((key) => {
      const value = sbType.value[key] as MetaDocgenInfo;
      return `
    â€¢${key}: ${value.type} ${value.description}
      `;
    })
    .join('<br>');
}
