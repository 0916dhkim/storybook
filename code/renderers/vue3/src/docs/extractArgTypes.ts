import type {
  SBEnumType,
  SBObjectType,
  SBScalarType,
  SBType,
  StrictArgTypes,
} from '@storybook/types';
import type { ArgTypesExtractor, DocgenInfo } from '@storybook/docs-tools';
import { hasDocgen, extractComponentProps } from '@storybook/docs-tools';

type Schema = { kind: string; schema: [] | object; type?: string } | string;
type MetaDocgenInfo = DocgenInfo & {
  type: string;
  default: string;
  global: boolean;
  name: string;
  schema: Schema;
  tags: { name: string; text: string }[];
};

const ARG_TYPE_SECTIONS = ['props', 'events', 'slots', 'exposed'];

export const extractArgTypes: ArgTypesExtractor = (component) => {
  if (!hasDocgen(component)) {
    return null;
  }

  const argTypes: StrictArgTypes = {};

  ARG_TYPE_SECTIONS.forEach((section) => {
    const props = extractComponentProps(component, section);

    props.forEach(({ docgenInfo }) => {
      const {
        name,
        type,
        description,
        default: defaultSummary,
        required,
        tags = [],
        global,
      } = docgenInfo as MetaDocgenInfo;

      if (argTypes[name] || global) {
        return; // skip duplicate and global props
      }

      const sbType = section === 'props' ? convert(docgenInfo as MetaDocgenInfo) : { name: 'void' };
      const nestedTypes =
        sbType.name === 'object' && section === 'props' ? nestedInfo(sbType as SBObjectType) : '';
      const definedTypes = `${type.replace(' | undefined', '')}`;
      const descriptions = `${
        tags.length ? `${tags.map((tag) => `@${tag.name}: ${tag.text}`).join('<br>')}<br><br>` : ''
      }${description} ${nestedTypes} `;

      argTypes[name] = {
        name,
        description: descriptions.replace('undefined', ''),
        defaultValue: { summary: defaultSummary },
        type: { required, ...sbType } as SBType,
        table: {
          type: { summary: definedTypes },
          jsDocTags: tags,
          defaultValue: { summary: defaultSummary },
          category: section,
        },
      };
    });
  });

  return argTypes;
};

export const convert = ({ schema: schemaType }: MetaDocgenInfo): SBType => {
  if (typeof schemaType !== 'object') {
    return { name: schemaType } as SBScalarType;
  }
  if (
    typeof schemaType === 'object' &&
    schemaType.kind === 'enum' &&
    Array.isArray(schemaType.schema)
  ) {
    const values = schemaType.schema
      .filter((item) => item !== 'undefined' && item !== null)
      .map((item) => (typeof item === 'string' ? item.replace(/"/g, '') : item));

    const sbType: SBType = { name: 'enum', value: values };
    const stringIndex = values.indexOf('string');
    const numberIndex = values.indexOf('number');
    const booleanIndex = values.indexOf('boolean');
    if (stringIndex !== -1 || numberIndex !== -1 || booleanIndex !== -1) {
      const typeName = values[stringIndex ?? numberIndex ?? booleanIndex];
      return { ...sbType, name: typeName, value: undefined } as SBScalarType;
    }
    const hasObject = values.find((item) => typeof item === 'object');

    return {
      ...sbType,
      name: hasObject ? 'array' : 'enum',
      value: hasObject ? undefined : values,
    } as SBEnumType;
  }
  if (
    typeof schemaType === 'object' &&
    schemaType.kind === 'object' &&
    typeof schemaType.schema === 'object'
  ) {
    const schemaObject = schemaType.schema as { [key: string]: MetaDocgenInfo };
    const props = Object.fromEntries(
      Object.entries(schemaObject).map(([key, value]) => {
        return [key, value as MetaDocgenInfo];
      })
    );
    return {
      name: 'object',
      value: props,
    } as SBObjectType;
  }
  return { name: schemaType } as unknown as SBType;
};

function nestedInfo(sbType: SBObjectType) {
  return Object.keys(sbType.value)
    .map((key) => {
      const value = sbType.value[key] as MetaDocgenInfo;
      return `
    â€¢${key}: ${value.type} ${value.description}
      `;
    })
    .join('<br>');
}
