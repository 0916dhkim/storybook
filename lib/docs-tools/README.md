# Storybook Docs Utils

Shared utility functions for frameworks to implement docs

- ArgType extraction
- Dynamic snippet generation
