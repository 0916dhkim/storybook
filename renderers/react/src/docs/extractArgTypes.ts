import type { StrictArgTypes } from '@storybook/csf';
import type { PropDef, ArgTypesExtractor } from '@storybook/docs-tools';
import { extractProps } from './extractProps';

export const extractArgTypes: ArgTypesExtractor = (component) => {
  if (component) {
    const extracted = extractProps(component);
    if (!extracted) {
      return null;
    }
    const { rows } = extracted;
    if (rows) {
      return rows.reduce((acc: StrictArgTypes, row: PropDef) => {
        const {
          name,
          description,
          type,
          sbType,
          defaultValue: defaultSummary,
          jsDocTags,
          required,
        } = row;

        acc[name] = {
          name,
          description,
          type: { required, ...sbType },
          table: {
            type,
            jsDocTags,
            defaultValue: defaultSummary,
          },
        };
        return acc;
      }, {});
    }
  }

  return null;
};
