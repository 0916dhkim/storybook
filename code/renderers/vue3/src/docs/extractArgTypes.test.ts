import { hasDocgen, extractComponentProps } from '@storybook/docs-tools';
import { extractArgTypes } from './extractArgTypes';
import {
  mockExtractComponentPropsReturn,
  referenceTypeProps,
  mockExtractComponentEventsReturn,
  referenceTypeEvents,
  templateSlots,
  mockExtractComponentSlotsReturn,
} from './tests-meta-components/meta-components';
import type { Mock } from 'vitest';
import { describe, expect, it, vitest } from 'vitest';

vitest.mock('@storybook/docs-tools');

describe('extractArgTypes', () => {
  it('should return null if component does not contain docs', () => {
    (hasDocgen as Mock).mockReturnValueOnce(false);
    (extractComponentProps as Mock).mockReturnValueOnce([] as any);

    expect(extractArgTypes({} as any)).toBeNull();
  });

  it('should extract arg types for component', () => {
    const component = referenceTypeProps;
    (hasDocgen as Mock).mockReturnValueOnce(true);
    (extractComponentProps as Mock).mockReturnValue(mockExtractComponentPropsReturn);

    const argTypes = extractArgTypes(component);

    expect(argTypes).toMatchInlineSnapshot(`
      {
        "array": {
          "control": {
            "disable": true,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "description required array object",
          "name": "array",
          "table": {
            "category": "events",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "MyNestedProps[]",
            },
          },
          "type": {
            "name": "MyNestedProps[]",
            "required": true,
          },
        },
        "arrayOptional": {
          "control": {
            "disable": true,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "description optional array object",
          "name": "arrayOptional",
          "table": {
            "category": "events",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "MyNestedProps[]",
            },
          },
          "type": {
            "name": "MyNestedProps[] | undefined",
            "required": false,
          },
        },
        "bar": {
          "control": {
            "disable": true,
          },
          "defaultValue": {
            "summary": "1",
          },
          "description": "description bar is optional number",
          "name": "bar",
          "table": {
            "category": "events",
            "defaultValue": {
              "summary": "1",
            },
            "jsDocTags": [],
            "type": {
              "summary": "number",
            },
          },
          "type": {
            "name": "number | undefined",
            "required": false,
          },
        },
        "baz": {
          "control": {
            "disable": true,
          },
          "defaultValue": {
            "summary": "["foo", "bar"]",
          },
          "description": "description baz is string array",
          "name": "baz",
          "table": {
            "category": "events",
            "defaultValue": {
              "summary": "["foo", "bar"]",
            },
            "jsDocTags": [],
            "type": {
              "summary": "string[]",
            },
          },
          "type": {
            "name": "string[] | undefined",
            "required": false,
          },
        },
        "enumValue": {
          "control": {
            "disable": true,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "description enum value",
          "name": "enumValue",
          "table": {
            "category": "events",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "MyEnum",
            },
          },
          "type": {
            "name": "MyEnum",
            "required": true,
          },
        },
        "foo": {
          "control": {
            "disable": true,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "@default: "rounded"<br>@since: v1.0.0<br>@see: https://vuejs.org/<br>@deprecated: v1.1.0<br><br>string foo",
          "name": "foo",
          "table": {
            "category": "events",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [
              {
                "name": "default",
                "text": ""rounded"",
              },
              {
                "name": "since",
                "text": "v1.0.0",
              },
              {
                "name": "see",
                "text": "https://vuejs.org/",
              },
              {
                "name": "deprecated",
                "text": "v1.1.0",
              },
            ],
            "type": {
              "summary": "string",
            },
          },
          "type": {
            "name": "string",
            "required": true,
          },
        },
        "inlined": {
          "control": {
            "disable": true,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "",
          "name": "inlined",
          "table": {
            "category": "events",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "{ foo: string; }",
            },
          },
          "type": {
            "name": "{ foo: string; }",
            "required": true,
          },
        },
        "literalFromContext": {
          "control": {
            "disable": true,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "description literal type alias that require context",
          "name": "literalFromContext",
          "table": {
            "category": "events",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": ""Uncategorized" | "Content" | "Interaction" | "Display" | "Forms" | "Addons"",
            },
          },
          "type": {
            "name": ""Uncategorized" | "Content" | "Interaction" | "Display" | "Forms" | "Addons"",
            "required": true,
          },
        },
        "nested": {
          "control": {
            "disable": true,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "description nested is required nested object",
          "name": "nested",
          "table": {
            "category": "events",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "MyNestedProps",
            },
          },
          "type": {
            "name": "MyNestedProps",
            "required": true,
          },
        },
        "nestedIntersection": {
          "control": {
            "disable": true,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "description required nested object with intersection",
          "name": "nestedIntersection",
          "table": {
            "category": "events",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "MyNestedProps & { additionalProp: string; }",
            },
          },
          "type": {
            "name": "MyNestedProps & { additionalProp: string; }",
            "required": true,
          },
        },
        "nestedOptional": {
          "control": {
            "disable": true,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "description optional nested object",
          "name": "nestedOptional",
          "table": {
            "category": "events",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "MyNestedProps | MyIgnoredNestedProps",
            },
          },
          "type": {
            "name": "MyNestedProps | MyIgnoredNestedProps | undefined",
            "required": false,
          },
        },
        "recursive": {
          "control": {
            "disable": true,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "",
          "name": "recursive",
          "table": {
            "category": "events",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "MyNestedRecursiveProps",
            },
          },
          "type": {
            "name": "MyNestedRecursiveProps",
            "required": true,
          },
        },
        "union": {
          "control": {
            "disable": true,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "description union is required union type",
          "name": "union",
          "table": {
            "category": "events",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "string | number",
            },
          },
          "type": {
            "name": "string | number",
            "required": true,
          },
        },
        "unionOptional": {
          "control": {
            "disable": true,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "description unionOptional is optional union type",
          "name": "unionOptional",
          "table": {
            "category": "events",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "string | number",
            },
          },
          "type": {
            "name": "string | number | undefined",
            "required": false,
          },
        },
      }
    `);
  });

  it('should extract events for Vue component', () => {
    const component = referenceTypeEvents;
    (hasDocgen as Mock).mockReturnValueOnce(true);
    (extractComponentProps as Mock).mockReturnValue(mockExtractComponentEventsReturn);

    const argTypes = extractArgTypes(component);

    expect(argTypes).toMatchSnapshot();
  });

  it('should extract slots type for Vue component', () => {
    const component = templateSlots;
    (hasDocgen as Mock).mockReturnValueOnce(true);
    (extractComponentProps as Mock).mockReturnValue(mockExtractComponentSlotsReturn);

    const argTypes = extractArgTypes(component);

    expect(argTypes).toMatchSnapshot();
  });
});
