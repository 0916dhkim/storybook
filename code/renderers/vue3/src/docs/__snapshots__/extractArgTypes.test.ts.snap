// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extractArgTypes > should extract events for Vue component 1`] = `
{
  "bar": {
    "control": {
      "disable": false,
    },
    "defaultValue": {
      "summary": undefined,
    },
    "description": "",
    "name": "bar",
    "table": {
      "category": "props",
      "defaultValue": {
        "summary": undefined,
      },
      "jsDocTags": [],
      "type": {
        "summary": "[value: { year: number; title?: any; }]",
      },
    },
    "type": {
      "name": [
        {
          "kind": "object",
          "schema": {
            "title": {
              "declarations": [
                {
                  "file": "/storybook/sandbox/vue3-vite-default-ts/src/stories/renderers/vue3/component-meta/reference-type-events/component.vue",
                  "range": [
                    269,
                    280,
                  ],
                },
              ],
              "description": "",
              "global": false,
              "name": "title",
              "required": false,
              "schema": "any",
              "tags": [],
              "type": "any",
            },
            "year": {
              "declarations": [
                {
                  "file": "/storybook/sandbox/vue3-vite-default-ts/src/stories/renderers/vue3/component-meta/reference-type-events/component.vue",
                  "range": [
                    255,
                    268,
                  ],
                },
              ],
              "description": "",
              "global": false,
              "name": "year",
              "required": true,
              "schema": "number",
              "tags": [],
              "type": "number",
            },
          },
          "type": "{ year: number; title?: any; }",
        },
      ],
      "required": undefined,
    },
  },
  "baz": {
    "control": {
      "disable": false,
    },
    "defaultValue": {
      "summary": undefined,
    },
    "description": "",
    "name": "baz",
    "table": {
      "category": "props",
      "defaultValue": {
        "summary": undefined,
      },
      "jsDocTags": [],
      "type": {
        "summary": "[]",
      },
    },
    "type": {
      "name": [],
      "required": undefined,
    },
  },
  "foo": {
    "control": {
      "disable": false,
    },
    "defaultValue": {
      "summary": undefined,
    },
    "description": "",
    "name": "foo",
    "table": {
      "category": "props",
      "defaultValue": {
        "summary": undefined,
      },
      "jsDocTags": [],
      "type": {
        "summary": "[data?: { foo: string; }]",
      },
    },
    "type": {
      "name": [
        {
          "kind": "enum",
          "schema": [
            "undefined",
            {
              "kind": "object",
              "schema": {
                "foo": {
                  "declarations": [
                    {
                      "file": "/storybook/sandbox/vue3-vite-default-ts/src/stories/renderers/vue3/component-meta/reference-type-events/component.vue",
                      "range": [
                        207,
                        218,
                      ],
                    },
                  ],
                  "description": "",
                  "global": false,
                  "name": "foo",
                  "required": true,
                  "schema": "string",
                  "tags": [],
                  "type": "string",
                },
              },
              "type": "{ foo: string; }",
            },
          ],
          "type": "{ foo: string; } | undefined",
        },
      ],
      "required": undefined,
    },
  },
}
`;

exports[`extractArgTypes > should extract slots type for Vue component 1`] = `
{
  "default": {
    "control": {
      "disable": false,
    },
    "defaultValue": {
      "summary": undefined,
    },
    "description": "",
    "name": "default",
    "table": {
      "category": "props",
      "defaultValue": {
        "summary": undefined,
      },
      "jsDocTags": [],
      "type": {
        "summary": "{ num: number; }",
      },
    },
    "type": {
      "name": "object",
      "required": undefined,
      "value": {
        "num": {
          "declarations": [],
          "description": "",
          "global": false,
          "name": "num",
          "required": true,
          "schema": "number",
          "tags": [],
          "type": "number",
        },
      },
    },
  },
  "named": {
    "control": {
      "disable": false,
    },
    "defaultValue": {
      "summary": undefined,
    },
    "description": "",
    "name": "named",
    "table": {
      "category": "props",
      "defaultValue": {
        "summary": undefined,
      },
      "jsDocTags": [],
      "type": {
        "summary": "{ str: string; }",
      },
    },
    "type": {
      "name": "object",
      "required": undefined,
      "value": {
        "str": {
          "declarations": [],
          "description": "",
          "global": false,
          "name": "str",
          "required": true,
          "schema": "string",
          "tags": [],
          "type": "string",
        },
      },
    },
  },
  "no-bind": {
    "control": {
      "disable": false,
    },
    "defaultValue": {
      "summary": undefined,
    },
    "description": "",
    "name": "no-bind",
    "table": {
      "category": "props",
      "defaultValue": {
        "summary": undefined,
      },
      "jsDocTags": [],
      "type": {
        "summary": "{}",
      },
    },
    "type": {
      "name": "object",
      "required": undefined,
      "value": {},
    },
  },
  "vbind": {
    "control": {
      "disable": false,
    },
    "defaultValue": {
      "summary": undefined,
    },
    "description": "",
    "name": "vbind",
    "table": {
      "category": "props",
      "defaultValue": {
        "summary": undefined,
      },
      "jsDocTags": [],
      "type": {
        "summary": "{ num: number; str: string; }",
      },
    },
    "type": {
      "name": "object",
      "required": undefined,
      "value": {
        "num": {
          "declarations": [
            {
              "file": "/storybook/sandbox/vue3-vite-default-ts/src/stories/renderers/vue3/component-meta/template-slots/component.vue",
              "range": [
                153,
                161,
              ],
            },
          ],
          "description": "",
          "global": false,
          "name": "num",
          "required": true,
          "schema": "number",
          "tags": [],
          "type": "number",
        },
        "str": {
          "declarations": [
            {
              "file": "/storybook/sandbox/vue3-vite-default-ts/src/stories/renderers/vue3/component-meta/template-slots/component.vue",
              "range": [
                163,
                173,
              ],
            },
          ],
          "description": "",
          "global": false,
          "name": "str",
          "required": true,
          "schema": "string",
          "tags": [],
          "type": "string",
        },
      },
    },
  },
}
`;
